tests/src/python/providertestbase.py

```diff
@@ -817,3 +816,24 @@ def testMinMaxAfterChanges(self):
             self.assertTrue(vl.dataProvider().deleteAttributes([0]))
             self.assertEqual(vl.dataProvider().minimumValue(0), -200)
             self.assertEqual(vl.dataProvider().maximumValue(0), 400)
+
+    def testStringComparison(self):
+        """
+        Test if string comparisons with numbers are cast by the expression
+        compiler (or work fine without doing anything :P)
+        """
+        for expression in (
+                '5 LIKE \'5\'',
+                '5 ILIKE \'5\'',
+                '15 NOT LIKE \'5\'',
+                '15 NOT ILIKE \'5\'',
+                '5 ~ \'5\''):
+            iterator = self.source.getFeatures(QgsFeatureRequest().setFilterExpression('5 LIKE \'5\''))
+            count = len([f for f in iterator])
+            self.assertEqual(count, 5)
+            self.assertFalse(iterator.compileFailed())
+            if self.enableCompiler():
+                iterator = self.source.getFeatures(QgsFeatureRequest().setFilterExpression('5 LIKE \'5\''))
+                self.assertEqual(count, 5)
+                self.assertFalse(iterator.compileFailed())
```

Comment:
Better check if the expression isn't in uncompiledFilters here - then we can indicate individual providers which don't handle this, add still catch regressions

Reply:
Given travis is happy, is this a requirement?

Reply:
Guess not... I'm just surprised no other providers have issues with this. I'd have thought some would. Maybe mssql/oracle which aren't tested on Travis (But who cares about those :P )

Reply:
Well, even if they suffer from this, the result will be correct (just slower). The test will only trigger once they start returning `true` for `compileFailed()`, right now they will silently fall back to local filtering.